sound/soc/sof/topology.c

```diff
@@ -1404,6 +1404,12 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 		return ret;
 	}
 
+	/* Fixup pin count in case they are not specified in topology */
+	if (!swidget->num_sink_pins)
+		swidget->num_sink_pins = 1;
+	if (!swidget->num_source_pins)
+		swidget->num_source_pins = 1;
+
```

Comment:
no that's not correct. It's perfectly legal to have a widget with no source or sink. We would need to have a different criterion to set the number to one.

Reply:
@aiChaoSONG can you please chime in on this issue?

Reply:
@plbossart, if it has no source or sink than how can it be connected? If it has only 1 source and/or sink then the number of source/sink is 1, not 0...

The other option I can think of is:
```diff
diff --git a/sound/soc/sof/ipc4-topology.c b/sound/soc/sof/ipc4-topology.c
index d2d61393a581..0de2ae0bfdb7 100644
--- a/sound/soc/sof/ipc4-topology.c
+++ b/sound/soc/sof/ipc4-topology.c
@@ -1608,15 +1608,8 @@ static int sof_ipc4_get_queue_id(struct snd_sof_widget *src_widget,
 
 	scomp = current_swidget->scomp;
 
-	if (num_pins < 1) {
-		dev_err(scomp->dev, "invalid %s num_pins: %d for queue allocation for %s\n",
-			(pin_type == SOF_PIN_TYPE_SOURCE ? "source" : "sink"),
-			num_pins, current_swidget->widget->name);
-		return -EINVAL;
-	}
-
 	/* If there is only one sink/source pin, queue id must be 0 */
-	if (num_pins == 1)
+	if (num_pins <= 1)
 		return 0;
 
 	/* Allocate queue ID from pin binding array if it is defined in topology. */
```

Reply:
The widget needs to have at least one sink or one source. Not both as your initial patch suggested it.

I am not sure the new version is better, but will let @aiChaoSONG reply. we don't want to allocate a queue, even with index zero, if there's no connection :-)

Reply:
If you have made your decision to use a fixup, another fix can be initializing
```
swidget->num_sink_pins = 1;
swidget->num_source_pins = 1;
```
around line 1394